### PR TITLE
Rename Perl version macro

### PR DESCRIPTION
**Pull request by atoomic**
This commit introduces Semantic Versioning macro for Perl version.

The previously named variables: PERL_REVISION, PERL_VERSION, PERL_SUBVERSION
are renamed as PERL_VERSION_MAJOR, PERL_VERSION_MINOR, and PERL_VERSION_PATCH

A new variable PERL_VERSION_HEX is also introduced.

References: https://semver.org

### CODE REVIEW

File: ext/Pod-Html/testdir/perlvar-copy.pod [hunk 1]
```diff
@@ -1331,10 +1331,10 @@ startup. This information is used by perl when it's in
 adjust-utf8ness-to-locale mode (as when run with the C<-CL> command-line
 switch); see L<perlrun> for more info on this.
 
-=item $PERL_VERSION
+=item $PERL_VERSION_MINOR
 
 =item $^V
-X<$^V> X<$PERL_VERSION>
+X<$^V> X<$PERL_VERSION_MINOR>
```
Comment (tonycoz) [inline ext/Pod-Html/testdir/perlvar-copy.pod]: This a) doesn't need to change (it's a test input file) and b) the names here are the perl variables exported by English.pm

**Review by tonycoz — CHANGES_REQUESTED** [review]
This doesn't really add semantic versioning, which is more a policy change which this change doesn't include (which would involve switching from our odd minor version == development series and other changes.)

The semvar spec doesn't discuss macros so I don't think the title of the PR, the commit message or the comment in patchlevel.h should refer to it.

I don't think this should be applied without wider discussion on the policy change.

File: ext/Pod-Html/testdir/perlvar-copy.pod (hunk 1 above)
Comment (atoomic) [reply ext/Pod-Html/testdir/perlvar-copy.pod]: indeed, this is now reverted

**Comment by atoomic** [comment]
@tonycoz I marked this PR as draft, as this need discussion and performed the several changes you requested via cfbc8487e41595e5c9e55b8f6b1e72111f03c0c0

**Comment by jkeenan** [comment]
@atoomic, can you provide an update on the status of this p.r. for semantic versioning?

Thank you very much.
Jim Keenan

**Comment by Leont** [comment]
There are several question here:

1. Should we add these symbols at all?
2. Should they be aliases for PERL_REVISION, PERL_VERSION, and PERL_SUBVERSION; or should it be the other way around.
3. That do they really mean. Should they be interpreted as strict semantic versioning, or as the current versioning scheme.

Unless the answer to the first question is no, all of these should be answered (though that answer may depend on the third question).

**Comment by jkeenan** [comment]
> There are several question here:
> 
>     1. Should we add these symbols at all?
> 
>     2. Should they be aliases for PERL_REVISION, PERL_VERSION, and PERL_SUBVERSION; or should it be the other way around.
> 
>     3. That do they really mean. Should they be interpreted as strict semantic versioning, or as the current versioning scheme.
> 
> 
> Unless the answer to the first question is no, all of these should be answered (though that answer may depend on the third question).

No one has responded to @Leont's questions in a year-and-a-half.  There's no reason to retain p.r.'s that are in Draft status.  I recommend that this p.r. be closed.

**Comment by tonycoz** [comment]
I'm pretty sure this change is obsolete, or way, way ahead of its time.

I think the intent was new symbols for a new Perl 7, while maintaining the old names for compatibility with code that only checked PERL_VERSION and PERL_SUBVERSION when checking for compatibility/new features.

**Comment by leonerd** [comment]
I'm not sure it's obsolete. There is still the intention to be able to bump the first component of the version number beyond 5 and into 7, 8, and beyond at *some* point in the future.

There are many obstacles to doing this, but one of them is the fact that so much existing code on CPAN has a `ppport.h` which insists on `PERL_REVISION == 5`.

The new macro form at least partly solves this, and does so by renaming the fields to what has been generally accepted industry-wide as a more standard naming structure of `MAJOR.MINOR.PATCH`. Overall it's a good idea, and I'm in favour of it.

**Comment by Leont** [comment]
> I'm not sure it's obsolete. There is still the intention to be able to bump the first component of the version number beyond 5 and into 7, 8, and beyond at _some_ point in the future.
> 
> There are many obstacles to doing this, but one of them is the fact that so much existing code on CPAN has a `ppport.h` which insists on `PERL_REVISION == 5`.
> 
> The new macro form at least partly solves this, and does so by renaming the fields to what has been generally accepted industry-wide as a more standard naming structure of `MAJOR.MINOR.PATCH`. Overall it's a good idea, and I'm in favour of it.

That makes sense, and answers at least the first two of my questions.

**Comment by jkeenan** [comment]
This merge request was created more than two years ago.  I reviewed it today.  @Leont indicates that 2 of his 3 questions have been answered, but that appears to leave this one:

*[What do @atoomic's new or renamed variables] really mean? Should they be interpreted as strict semantic versioning, or as the current versioning scheme?*

Could we focus discussion on that?

Also: @atoomic, this p.r. has acquired many merge conflicts.  Could you clear them up?

Thank you very much.
Jim Keenan

**Comment by haarg** [comment]
To clarify something, these symbols already exist as aliases to the classic version defines. This PR changes the rest of the code to use the new symbols, and switches the classic defines to be aliases for the new ones.

Even without any change in policy, I think this is a good change to make. I think it's reasonable to refer to the different components of the version as MAJOR, MINOR, and PATCH under the current versioning scheme. And switching everything to use these could allow the old defines to be given "fake" values to assist in bumping the major version past 5.

A policy change related to semver should be considered separately. I doubt we're anywhere close to wanting to make that change. Generally semver is interpreted to also apply to binary compatibility. This would mean bumping the major version each year under our current development process. I don't think this is generally what people want.

File: patchlevel.h [hunk 2]
```diff
@@ -39,9 +39,26 @@ Instead use one of the version comparison macros.  See C<L</PERL_VERSION_EQ>>.
...
+       - PERL_VERSION_LT
+       - PERL_VERSION_GT
+       - PERL_VERSION_LE
+       - PERL_VERSION_GE
+*/
+#define PERL_VERSION_MAJOR        5
+#define PERL_VERSION_MINOR       33
+#define PERL_VERSION_PATCH        1
+#define PERL_VERSION_RELEASE      0 /* for future usage */
+#define PERL_VERSION_BUILD        0 /* for future usage */
```
Comment (leonerd) [inline patchlevel.h:54]: I'm cautious about defining these new "for future use" ones, without having a clearer plan of what possible future use they might have.

File: t/run/switches.t [hunk 3]
```diff
@@ -301,8 +301,8 @@ is runperl(stderr => 1, prog => '#!perl -M'),
         skip "Win32 miniperl produces a default archname in -v", 1
 	  if $^O eq 'MSWin32' && is_miniperl;
         my $v = sprintf "%vd", $^V;
-        my $ver = $Config{PERL_VERSION};
-        my $rel = $Config{PERL_SUBVERSION};
+        my $ver = $Config{PERL_VERSION_MINOR};
+        my $rel = $Config{PERL_VERSION_PATCH};
```
Comment (leonerd) [inline t/run/switches.t:311]: Probably makes sense to rename these to `$ver_minor` and `$ver_patch` for consistency

**Review by leonerd — CHANGES_REQUESTED** [review]
I'd suggest removing those two "reserved for future use" ones for now, so this becomes a simple change of the names of the existing variables, with no wider considerations of policy. We can consider other changes later once this is in.

**Comment by leonerd** [comment]
I'm definitely keen to see this go in. The new variable names are neater and easier to remember (major.minor.patch is industry-standard, as compared our previous revision.version.subversion - what even is that??!) and also have a certain visual neatness of all being the same length so for *that reason alone* it's already an improvement.

Additionally it would give us the flexibility to bump the major version up to 7 or beyond at some future date, which is a change we currently cannot make. So keeping that in mind this is also good.

I also feel this is purely a technical-mechanism change. We're not saying "adopt semver" with all the policy suggestions that makes. The code and the commit message don't mention semver, but this PR description currently still does. As long as we all accept that **this change itself does not introduce semver** I don't think that matters.. but maybe you want to edit the PR description as well to remove that mention if the discussion continues.

But I don't think it has to. The principle seems sound, I have no unanswered questions. If you fix up the conflicts and generally make this merge-ready I'd be happy to give it a review in the hope of merging it and finally putting this one to bed.

**Comment by Leont** [comment]
> The new variable names are neater and easier to remember

I'm not sure that is relevant, given that everyone really ought to be using `PERL_VERSION_GE` and `PERL_VERSION_LT` anyway.

**Comment by atoomic** [comment]
I can rebase and rework that Pull Request but I will only do that if there is any interest in merging this or part of this Pull Request.
If the consensus is that this does not provide value, then let's close the pull request.

**Comment by khwilliamson** [comment]
On 9/19/22 08:55, ℕicolas ℝ. wrote:
> I can rebase and rework that Pull Request but I will only do that if 
> there is any interest in merging this or part of this Pull Request.
> If the consensus is that this does not provide value, then let's close 
> the pull request.

I would like this done, so +1

**Comment by demerphq** [comment]
On Mon, 19 Sept 2022, 22:31 Karl Williamson, ***@***.***>
wrote:

> On 9/19/22 08:55, ℕicolas ℝ. wrote:
> > I can rebase and rework that Pull Request but I will only do that if
> > there is any interest in merging this or part of this Pull Request.
> > If the consensus is that this does not provide value, then let's close
> > the pull request.
>
> I would like this done, so +1
>

Karl, t is not entirely clear what "this" is. I assume you want it merged,
but your statement supports both interpretations.

Yves

Yves

>

**Comment by khwilliamson** [comment]
On 9/19/22 16:08, Yves Orton wrote:
> On Mon, 19 Sept 2022, 22:31 Karl Williamson, ***@***.***>
> wrote:
> 
>  > On 9/19/22 08:55, ℕicolas ℝ. wrote:
>  > > I can rebase and rework that Pull Request but I will only do that if
>  > > there is any interest in merging this or part of this Pull Request.
>  > > If the consensus is that this does not provide value, then let's close
>  > > the pull request.
>  >
>  > I would like this done, so +1
>  >
> 
> Karl, t is not entirely clear what "this" is. I assume you want it merged,
> but your statement supports both interpretations.
> 

Sorry.  I think this PR is a step forward, and should be merged when it 
is ready to be so.

File: Porting/Glossary [hunk 4]
```diff
@@ -4715,6 +4715,18 @@ PERL_VERSION (Oldsyms.U):
...
 
+PERL_VERSION_MAJOR (Oldsyms.U):
+	In a Perl version number such as X.Y.Z, this is the X.
+	This value is manually set in patchlevel.h
+
+PERL_VERSION_MINOR (Oldsyms.U):
+	In a Perl version number such as X.Y.Z, this is the Y.
+	This value is manually set in patchlevel.h
+
+PERL_VERSION_RELEASE (Oldsyms.U):
```
Comment (haarg) [inline Porting/Glossary]: Should this be `PERL_VERSION_PATCH`?

Comment (atoomic) [reply Porting/Glossary]: true but we probably should not even edit that file by hand,`./metaconfig.SH` is in charge of it
going to revert the changes from this file

File: configpm [hunk 5]
```diff
@@ -225,6 +225,20 @@ i_time='define'
 prototype='define'
 EOT
 
+# Don't lie about PATCHLEVEL and SUBVERSION in %Config
```
Comment (leonerd) [inline configpm:228]: I'm not sure I get what this comment is about. What kind of lying would happen if we didn't do these lines of code?

**Review by leonerd — COMMENTED** [review]
Added one more comment. There's still some of mine from 17th Sept that haven't been addressed yet.

**Comment by demerphq** [comment]
@atoomic the consensus seems to be that people *do* want to merge this, but that there are some minor changes required before it merges. Also the PR is currently in conflict. Do you want to get it up to date or should someone else pick it up?

**Comment by atoomic** [comment]
No problems @demerphq I will revisit it an extra time :-) and submit and updated PR

**Comment by jkeenan** [comment]
> No problems @demerphq I will revisit it an extra time :-) and submit and updated PR

@atoomic, are we going forward with this pull request?  (It now has merge conflicts.)  Thanks.

**Review by leonerd — COMMENTED** [review]
It'd be good to get this tidied up and ready to go. Can someone - @atoomic or elsewise - rebase and tidy it up ready to merge?

**Comment by jkeenan** [comment]
> It'd be good to get this tidied up and ready to go. Can someone - @atoomic or elsewise - rebase and tidy it up ready to merge?

In the `jkeenan/semantic-versioning` branch in the main repository, I have *attempted* to rebase @atoomic's branch on blead.

https://github.com/Perl/perl5/tree/jkeenan/semantic-versioning

https://github.com/Perl/perl5/compare/jkeenan/semantic-versioning?expand=1

Since it took me 4 tries to get it right, please review it carefully.  I'm not sure how or whether I make that code the content of this pull request.  Please advise.

**Comment by jkeenan** [comment]
> > It'd be good to get this tidied up and ready to go. Can someone - @atoomic or elsewise - rebase and tidy it up ready to merge?
> 
> In the `jkeenan/semantic-versioning` branch in the main repository, I have _attempted_ to rebase @atoomic's branch on blead.
> 
> https://github.com/Perl/perl5/tree/jkeenan/semantic-versioning
> 
> https://github.com/Perl/perl5/compare/jkeenan/semantic-versioning?expand=1
> 
> Since it took me 4 tries to get it right, please review it carefully. I'm not sure how or whether I make that code the content of this pull request. Please advise.

@atoomic @leonerd can you please review above?  Thanks.

**Comment by jkeenan** [comment]
> > It'd be good to get this tidied up and ready to go. Can someone - @atoomic or elsewise - rebase and tidy it up ready to merge?
> 
> In the `jkeenan/semantic-versioning` branch in the main repository, I have _attempted_ to rebase @atoomic's branch on blead.
> 
> https://github.com/Perl/perl5/tree/jkeenan/semantic-versioning
> 
> https://github.com/Perl/perl5/compare/jkeenan/semantic-versioning?expand=1
> 
> Since it took me 4 tries to get it right, please review it carefully. I'm not sure how or whether I make that code the content of this pull request. Please advise.

My branch which attempted to rebase @atoomic's original p.r. on blead has now been updated here:
https://github.com/Perl/perl5/compare/blead...jkeenan:perl5:semantic-version-upgrade-20240324?expand=1

**Comment by jkeenan** [comment]
This pull request has had unresolved merge conflicts for approximately a year-and-a-half.  If someone really believes we should go forward with these variable renamings, they are welcome to give it a fresh start in a new pull request.  In the meantime, closing this ticket.

**Pull request closed without merging**